File: 프로그래머스/4/42897. 도둑질/도둑질.py
def solution(money):
    # 첫 집을 무조건 터는 경우 -> 맨 마지막은 털지 못한다.
    DP1 = [0] * len(money)
    DP1[0] = money[0]
    DP1[1] = max(money[0], money[1])
    
    for i in range(2, len(money)-1):
        DP1[i] = max(DP1[i-1], DP1[i-2]+money[i])
        
    # 마지막 집을 무조건 터는 경우 -> 첫 집은 털지 못한다.
    DP2 = [0] * len(money)
    DP2[1] = money[1]
    
    for i in range(2, len(money)):
        DP2[i] = max(DP2[i-1], DP2[i-2]+money[i])
    
    return max(max(DP1), max(DP2))
        

File: 프로그래머스/4/42897. 도둑질/도둑질.java
import java.io.*;
import java.util.*;

class Solution {
    
    public int calculate(int[] money, int start, int end) {
        
        int prev1 = 0;  // i-1 집
        int prev2 = 0;  // i-2 집
        
        for ( int i = start ; i <= end ; i++ ) {
            int cur = Math.max(prev2 + money[i], prev1);
            prev2 = prev1;
            prev1 = cur;
        }
        
        return prev1;
        
    }
    
    public int solution(int[] money) {
        
        int n = money.length;
        
        int firstStolen = calculate(money, 0, n-2);
        int firstNotStolen = calculate(money, 1, n-1);
        
        return Math.max(firstStolen, firstNotStolen);
        
    }
}